crates/hir/src/semantics.rs

```diff
@@ -87,6 +87,28 @@ impl PathResolution {
     }
 }
 
+#[derive(Debug)]
+pub struct TypeInfo {
+    /// The original type of the expression or pattern.
+    pub ty: Type,
+    /// The coerced type, if a coercion happened.
+    pub coerced: Option<Type>,
+}
+
+impl TypeInfo {
+    pub fn ty(self) -> Type {
+        self.ty
+    }
+
+    pub fn coerced(self) -> Option<Type> {
+        self.coerced
+    }
+
+    pub fn coerced_or_original(self) -> Type {
+        self.coerced.unwrap_or(self.ty)
+    }
```

Comment:
Thinking more about the API, it feels that every expression always has an effective type, and unadjusted type, which are never optional. It's just that most of the time the two happen to coincide. 

So perhaps something like this:

```rust
impl TypeInfo {
    fn ty() -> Type;
    fn coerced() -> Type;
    fn has_coercion() -> bool;
}
```

not sure whether coerced or unccoerced should be the default. I think that most of the time you want coerced one actualy?

Reply:
Ye, that seems like a good choice.

I think in most cases you want to use the coerced one if possible except for a few cases. I've been pretty conservative with using the coerced type though. I'll do another round of checks where using the coerced type would make sense.

Reply:
My feeling is that unadjusted should be the default. It's the less surprising type (what's the type of `&&&&foo`? `&&&&T` or `&T`?) and for a long time was the only one we had...

Reply:
Or we just dodge the question by calling them `original` and `coerced` (or `adjusted`, since not all adjustments come from coercion...)

Reply:
+1 for `original` and `ajudsted` names from me. This seems clearest to me.

Reply:
No strong opinion!